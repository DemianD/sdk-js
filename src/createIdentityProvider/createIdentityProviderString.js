function encodeParams(options = {}) {
  return Object.keys(options)
    .map(key => `${key}=${encodeURIComponent(options[key])}`)
    .join('&')
}

export default function createIdentityProviderString(type, options) {
  return `${type}://${encodeParams(options)}`
}
