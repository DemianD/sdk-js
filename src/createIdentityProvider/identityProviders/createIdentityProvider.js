import retry from '../../util/retry'

import { createIdentityProviderError } from '../../util/createError'

import {
  MISCONFIGURED,
  MISCONFIGURED_REALM,
  MISCONFIGURED_AUTH_PROVIDER_OPTIONS,
} from '../../constants'
import { joinAsSpeech, AND } from '../../util/joinAsSpeech'
import omit from '../../util/omit'

const MAX_ATTEMPTS = 3
const REFETCH_WINDOW = 20

function secondsToMilliseconds(seconds) {
  return seconds * 1000
}

export function createIdentityProvider(
  grant_type,
  requiredParameters = [],
  optionalParameters = []
) {
  return class {
    constructor(network, parameters = {}) {
      this._network = network
      this._parameters = parameters

      // Check if we got a realm
      const { realm } = parameters
      if (!realm) {
        throw createIdentityProviderError(
          `realm is a required option for \`${grant_type}\``,
          MISCONFIGURED,
          MISCONFIGURED_REALM
        )
      }

      const allPossibleParameters = [
        ...requiredParameters,
        ...optionalParameters,
      ]

      const givenParameters = Object.keys(parameters).filter(
        key => key !== 'realm' // Skip the realm option, will be validated by its own
      )

      const missingRequiredParameters = requiredParameters.filter(
        required => !givenParameters.includes(required)
      )

      const extraMessage = `\n  - Required parameters: ${joinAsSpeech(
        AND,
        requiredParameters
      ) || '<none>'}\n  - Optional parameters: ${joinAsSpeech(
        AND,
        optionalParameters
      ) || '<none>'}`

      if (missingRequiredParameters.length > 0) {
        throw createIdentityProviderError(
          `${joinAsSpeech(AND, missingRequiredParameters)} ${
            missingRequiredParameters.length === 1
              ? 'is a required parameter'
              : 'are required parameters'
          } for \`${grant_type}\`${extraMessage}`,
          MISCONFIGURED,
          MISCONFIGURED_AUTH_PROVIDER_OPTIONS
        )
      }

      const unnecessaryParameters = givenParameters.filter(
        parameter => !allPossibleParameters.includes(parameter)
      )

      if (unnecessaryParameters.length > 0) {
        throw createIdentityProviderError(
          `${joinAsSpeech(AND, unnecessaryParameters)} ${
            unnecessaryParameters.length === 1
              ? 'is not an optional or a required parameter'
              : 'are not optional or required parameters'
          } for \`${grant_type}\`${extraMessage}`,
          MISCONFIGURED,
          MISCONFIGURED_AUTH_PROVIDER_OPTIONS
        )
      }
    }

    getAuthorization() {
      const { realm } = this._parameters

      if (this._current === undefined) {
        this._current = retry(
          (resolve, reject) => {
            // Get the access_token
            this._network
              .post(
                `${realm}/access_tokens`,
                Object.assign({}, omit(this._parameters, ['realm']), {
                  grant_type,
                })
              )
              .then(resolve, reject)
          },
          {
            max_attempts: MAX_ATTEMPTS,
          }
        )
          // Setup all the required parts to authenticate
          .then(({ data }) => ({
            Authorization: `${data.token_type} ${data.access_token}`,
            Expiration: data.expires_in,
            Realm: realm,
          }))

          // Get the proxy url
          .then(({ Realm, Authorization, Expiration }) =>
            this._network
              .get(`${Realm}/integrations/proxy`, {
                headers: { Authorization },
              })
              .then(({ data: response }) => ({
                Realm: response.data.url,
                Authorization,
                Expiration,
              }))
          )

          // Make sure to create a new access_token when the current one is going to expire
          .then(access => {
            setTimeout(() => {
              this._current = undefined
              this.getAuthorization()
            }, secondsToMilliseconds(access.Expiration - REFETCH_WINDOW))
            return access
          })
      }

      return this._current
    }
  }
}
