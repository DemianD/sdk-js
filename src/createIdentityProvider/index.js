import createError from '../util/createError'
import { MISCONFIGURED, MISCONFIGURED_AUTH_PROVIDER } from '../constants'
import joinAsSpeech, { AND } from '../util/joinAsSpeech'

import Client from './Client'
import parse from './parseIdentityProviderString'

const IDPS = {
  client: Client,
}

export default function createIdentityProvider(idp_descriptor, idps = IDPS) {
  const { type, options } = parse(idp_descriptor)

  if (idps[type] === undefined) {
    const providers = Object.keys(idps).map(key => `\`${key}\``)

    throw createError(
      `Identity provider \`${type}\` does not exist. The only ${
        providers.length === 1 ? `option is` : `options are`
      } ${joinAsSpeech(AND, providers)}.`,
      MISCONFIGURED,
      MISCONFIGURED_AUTH_PROVIDER
    )
  }

  return new idps[type](options)
}
