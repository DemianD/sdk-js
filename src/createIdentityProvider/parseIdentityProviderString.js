import { createIdentityProviderError } from '../util/createError'

import { MISCONFIGURED, MISCONFIGURED_AUTH_PROVIDER } from '../constants'

export default function parseIdentityProviderString(idp_descriptor = '') {
  const [type, encoded_options] = idp_descriptor.split('://')

  if (type === undefined || type === '') {
    throw createIdentityProviderError(
      'You should provide a `type` as identity provider.',
      MISCONFIGURED,
      MISCONFIGURED_AUTH_PROVIDER
    )
  }

  if (encoded_options === undefined || encoded_options === '') {
    throw createIdentityProviderError(
      'You should provide `options` for the identity provider.',
      MISCONFIGURED,
      MISCONFIGURED_AUTH_PROVIDER
    )
  }

  const options = encoded_options
    .split('&')
    .map(item => item.split('='))
    .map(([key, value]) => ({
      [key]: decodeURIComponent(value),
    }))
    .reduce((a, b) => Object.assign(a, b))

  return { type, options }
}
