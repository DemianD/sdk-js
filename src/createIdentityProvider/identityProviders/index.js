import { createIdentityProvider } from './createIdentityProvider'
import Token from './Token'

export const IDPS = {
  client_credentials: createIdentityProvider(
    'client_credentials',
    ['client_id', 'client_secret'],
    ['scope']
  ),
  office365_code: createIdentityProvider('office365_code', [
    'code',
    'client_id',
  ]),
  password: createIdentityProvider(
    'password',
    ['client_id', 'client_secret', 'username', 'password'],
    ['scope', 'enterprise_id']
  ),
  public_client: createIdentityProvider(
    'public_client',
    ['client_id'],
    ['scope']
  ),
  public_client_password: createIdentityProvider(
    'public_client_password',
    ['client_id', 'username', 'password'],
    ['scope', 'enterprise_id']
  ),
  public_client_password_reset: createIdentityProvider(
    'public_client_password_reset',
    ['client_id', 'username'],
    ['password']
  ),
  resource_code: createIdentityProvider('resource_code', ['client_id', 'code']),
  saml: createIdentityProvider('saml', [
    'payload',
    'enterprise_id',
    'client_id',
  ]),

  token: Token,
}
