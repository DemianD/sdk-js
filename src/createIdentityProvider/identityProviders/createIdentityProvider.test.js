import { createIdentityProvider } from './createIdentityProvider'

describe('createIdentityProvider', () => {
  const network = {} // Let's mock the network for now

  it('should error when no parameters are provided', () => {
    const MyIdentityProvider = createIdentityProvider('my_identity_provider')

    expect(() => new MyIdentityProvider()).toThrowErrorMatchingSnapshot()
  })

  it('should be possible to create an identity provider without required or optional parameters', () => {
    const MyIdentityProvider = createIdentityProvider('my_identity_provider')
    const myIdentityProviderInstance = new MyIdentityProvider(network, {
      realm: 'some-realm', // A realm is required though
    })

    expect(myIdentityProviderInstance).toBeInstanceOf(MyIdentityProvider)
  })

  it('should be possible to create an identity provider without required or optional parameters, error on an unnecessary parameter', () => {
    const MyIdentityProvider = createIdentityProvider('my_identity_provider')

    expect(
      () =>
        new MyIdentityProvider(network, {
          realm: 'some-realm', // A realm is required though
          unnecessary_key: 'unnecessary_value',
        })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should be possible to create an identity provider without required or optional parameters, error on multiple unnecessary parameters', () => {
    const MyIdentityProvider = createIdentityProvider('my_identity_provider')

    expect(
      () =>
        new MyIdentityProvider(network, {
          realm: 'some-realm', // A realm is required though
          unnecessary_key_1: 'unnecessary_value_1',
          unnecessary_key_2: 'unnecessary_value_2',
        })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should error when a required parameter is missing', () => {
    const MyIdentityProvider = createIdentityProvider('my_identity_provider', [
      'required_a',
      'required_b',
    ])

    expect(
      () =>
        new MyIdentityProvider(network, {
          realm: 'some-realm',
          required_a: 'required a value', // Note that required_b is missing
        })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should error when muliptle required parameters are missing', () => {
    const MyIdentityProvider = createIdentityProvider('my_identity_provider', [
      'required_a',
      'required_b',
      'required_c',
    ])

    expect(
      () =>
        new MyIdentityProvider(network, {
          realm: 'some-realm',
          required_a: 'required a value', // Note that required_b and required_c are missing
        })
    ).toThrowErrorMatchingSnapshot()
  })
})
