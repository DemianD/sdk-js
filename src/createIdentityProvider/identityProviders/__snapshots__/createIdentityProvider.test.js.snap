// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIdentityProvider should be possible to create an identity provider without required or optional parameters, error on an unnecessary parameter 1`] = `
"[IDENTITY PROVIDER]: unnecessary_key is not an optional or a required parameter for \`my_identity_provider\`
  - Required parameters: <none>
  - Optional parameters: <none>"
`;

exports[`createIdentityProvider should be possible to create an identity provider without required or optional parameters, error on multiple unnecessary parameters 1`] = `
"[IDENTITY PROVIDER]: unnecessary_key_1 and unnecessary_key_2 are not optional or required parameters for \`my_identity_provider\`
  - Required parameters: <none>
  - Optional parameters: <none>"
`;

exports[`createIdentityProvider should error when a required parameter is missing 1`] = `
"[IDENTITY PROVIDER]: required_b is a required parameter for \`my_identity_provider\`
  - Required parameters: required_a and required_b
  - Optional parameters: <none>"
`;

exports[`createIdentityProvider should error when muliptle required parameters are missing 1`] = `
"[IDENTITY PROVIDER]: required_b and required_c are required parameters for \`my_identity_provider\`
  - Required parameters: required_a, required_b and required_c
  - Optional parameters: <none>"
`;

exports[`createIdentityProvider should error when no parameters are provided 1`] = `"[IDENTITY PROVIDER]: realm is a required option for \`my_identity_provider\`"`;
